### DISCUSSION TITLE
Can startup time be improved to be on par with other languages?

### DISCUSSION THREAD

**Discussion by hlandau** [racket/rhombus#331] [Ideas]
I'd like to get into Racket as a Scheme user, but its high startup times are a blocker for me:

  $ time racket -e '(display "hello")'
  hello
  real 0m0.721s
  user 0m0.598s
  sys 0m0.119s

This basically rules out using it for command line scripts, which is a major class of uses for me.

Some have suggested using `#lang racket/base` and `raco make` to reduce the time. This reduces execution time to 0.175s. However this is still worse than Python and Guile, which manage 0.033s and 0.022s respectively, neither of which need a precompile step (which would make this kind of usage impractical).

Could this be improved in a future Racket?

**Comment by mfelleisen** [comment]
Rash? https://docs.racket-lang.org/rash/

**Comment by soegaard** [comment]
> Could this be improved in a future Racket?

It's a problem we are aware of, but the low-hanging fruit have been picked.

On my laptop from 2019 the numbers are:

  0.48s user 
  0.22s system 67% cpu 
  1.042 total

I expect newer systems to be faster.

Note that the startup time will be neglible when the script actually has work to do.

If you find the problem large enough, a potential solution, you could explore:

  - make a `racketd` that runs in the background 
  - let your script invoke `racketd`.
  
This solution is/was used by Emacs: 
    https://www.emacswiki.org/emacs/EmacsAsDaemon

Would it be ok to close this again?

**Comment by mflatt** [comment]
One of the obstacles to a fast startup time in Racket is that a lot of things are expected to be available in all contexts, like `eval` or individual modules that are part of the `racket/base` module. We may be able to reset those expectations and guarantees so that a `rhombus` executable can start faster, but it's an open problem at this point.

**Comment by LiberalArtist** [comment]
> I'd like to get into Racket as a Scheme user, but its high startup times are a blocker for me:
> 
> $ time racket -e '(display "hello")' hello real 0m0.721s user 0m0.598s sys 0m0.119s
> 
> This basically rules out using it for command line scripts, which is a major class of uses for me.
> 
> Some have suggested using `#lang racket/base` and `raco make` to reduce the time. This reduces execution time to 0.175s. However this is still worse than Python and Guile, which manage 0.033s and 0.022s respectively, neither of which need a precompile step (which would make this kind of usage impractical).
> 
> Could this be improved in a future Racket?

One difference to note when running code from a file is that Guile has auto-compilation enabled by default, which is equivalent to using the new-ish `-y`/`--make` flag to `racket`. You can put this in a shebang line with `#!/usr/bin/env -S racket -y` or similar.

More broadly, the Guile developers have put particular effort into startup time, and it has paid off. Nonetheless, many of us do write scripts in Racket!

I was surprised when I tried this that passing a program with `-e` took almost three times as long as running it from a (module) file, though:

```
philip@bastet:/tmp/times$ time racket -e '(display "hello")'
hello
real    0m0.231s
user    0m0.171s
sys     0m0.060s
hellophilip@bastet:/tmp/times$ echo '#lang racket/base (displayln "hello")' >hi.rkt
philip@bastet:/tmp/times$ time racket hi.rkt 
hello

real    0m0.079s
user    0m0.047s
sys     0m0.031s
philip@bastet:/tmp/times$ time racket -y hi.rkt 
hello

real    0m0.122s
user    0m0.073s
sys     0m0.035s
philip@bastet:/tmp/times$ time racket -y hi.rkt 
hello

real    0m0.097s
user    0m0.076s
sys     0m0.021s
philip@bastet:/tmp/times$ time racket hi.rkt 
hello

real    0m0.065s
user    0m0.041s
sys     0m0.024s
philip@bastet:/tmp/times$ time guile -c '(display "hello")'
hello
real    0m0.011s
user    0m0.006s
sys     0m0.006s
philip@bastet:/tmp/times$ echo '(display "hello")(newline)' >hi.scm
philip@bastet:/tmp/times$ time guile --no-auto-compile hi.scm 
hello

real    0m0.016s
user    0m0.009s
sys     0m0.009s
philip@bastet:/tmp/times$ time guile hi.scm 
;;; note: auto-compilation is enabled, set GUILE_AUTO_COMPILE=0
;;;       or pass the --no-auto-compile argument to disable.
;;; compiling /tmp/times/hi.scm
;;; compiled /home/philip/.cache/guile/ccache/3.0-LE-8-4.6/tmp/times/hi.scm.go
hello

real    0m0.035s
user    0m0.030s
sys     0m0.024s
philip@bastet:/tmp/times$ time guile hi.scm 
hello

real    0m0.011s
user    0m0.009s
sys     0m0.005s
```

(My `/tmp` is a tmpfs, and I ran the initial command a few times to make sure the filesystem cache was hot.)
